nova/virt/vmwareapi/special_spawning.py

```diff
@@ -306,9 +307,17 @@ def free_host(self, context):
                             'maintenance or became disconnected.')
                 return FREE_HOST_STATE_ERROR
 
+        # filter the VMs on the host, so we don't look at the non-movable
+        # DRS-created and -owned VMs
+        vcls_identifier = (constants.VCLS_EXTENSION_KEY,
+                           constants.VCLS_EXTENSION_TYPE_AGENT)
+        non_vcls_vms = [
+            (u, state) for u, h, state, m in self._get_vms_on_host(host_ref)
+            if (m.extensionKey, m.type) != vcls_identifier]
+
         # check if there are running VMs on that host
-        running_vms = [u for u, h, state in self._get_vms_on_host(host_ref)
-                        if state != 'poweredOff']
+        running_vms = [u for u, state in non_vcls_vms
+                       if state != 'poweredOff']
```

Comment:
What about merging this with the above `non_vcls_vms` logic, into 1 single function, that would be used to filter the `_get_vms_on_host()` in a single `runninng_vms` list ?

Reply:
The whole thin is only used once, I don't see a benefit in putting it into a function.

If you mean merging it into the same list-comprehension: I don't like that. I'd rather add one filter after another to keep a better overview in the code.